Prevent staking to a retired pool.

jonathanknowles opened the pull request:
# Issue Number

#1853

# Overview

This PR:

- [x] Adjusts `joinStakePool` to return `ErrNoSuchPool` if the current epoch is _later than or equal to_ the retirement epoch of the specified pool.

# Comments

Integration tests specifically designed to test the rejection mechanism are **not** included in this PR, but will be tackled in a future PR. (See QA section of issue https://github.com/input-output-hk/cardano-wallet/issues/1853.)

However, existing tests already verify that it is possible to stake to a _non-retired_ pool, so we can be fairly confident we are not breaking existing functionality.

jonathanknowles committed:
Reformat code to comply with coding standards.

jonathanknowles committed:
Add `ErrPoolAlreadyRetired` error.

jonathanknowles committed:
Add `Arbitrary` instance for `PoolRetirementEpochInfo`.

jonathanknowles committed:
Add `PoolRetirementEpochInfo` argument to `prop_guardJoinQuit`.

jonathanknowles committed:
Add `PoolRetirementEpochInfo` type.

jonathanknowles committed:
Add `PoolRetirementEpochInfo` argument to `guardJoinPool`.

jonathanknowles committed:
Enhance `ErrPoolAlreadyRetired` case of `prop_guardJoinQuit`.

jonathanknowles committed:
Enhance success case of `prop_guardJoinQuit`.

jonathanknowles committed:
Add coverage check to `prop_guardJoinQuit`.

jonathanknowles committed:
Use consistent argument order for `joinStakePool` functions.

jonathanknowles committed:
Rename argument to `knownPools`.

Justification: it's useful for the reader to know that this value refers
to all known pools, rather than some arbitrary subset.

jonathanknowles committed:
Add `getPoolLifeCycleStatus` operation to `StakePoolLayer`.

jonathanknowles committed:
Add `getPoolStatus` parameter to `Api.Server.joinStakePool`.

jonathanknowles committed:
Add `PoolLifeCycleStatus` parameter to `Wallet.joinStakePool`.

jonathanknowles committed:
Add `NetworkParameters` to `Wallet.joinStakePool`.

jonathanknowles committed:
Swap the field order for `PoolRetirementEpochInfo`.

jonathanknowles committed:
Use real retirement info in `Wallet.joinStakePool`.

jonathanknowles committed:
Avoid lambda in function `getPoolLifeCycleStatus`.

jonathanknowles committed:
Follow `hlint` suggestion to use `Just True ==`.

In function `prop_guardJoinQuit`.

jonathanknowles committed:
Encode more information in `ErrPoolAlreadyRetired`.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1715,36 +1718,53 @@ joinStakePool
-    let action = if isKeyReg then Join pid else RegisterKeyAndJoin pid
-    liftIO $ traceWith tr $ MsgIsStakeKeyRegistered isKeyReg
+        currentTime <- liftIO getCurrentTime
+        let currentEpoch = view #epochNumber
```
Could we please rebase this off #1901 ? 🙏 

Or, actually, no it's probably better to merge this first.

rvl reviewed (commented):
Is there a reason that we track retired pools? I don't think we would need info about pools which can't be delegated to. Shouldn't we purge them from the database, and return "pool not found" in this case?

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1726,36 +1729,53 @@ joinStakePool
+        currentTime <- liftIO getCurrentTime
+        let currentEpoch = view #epochNumber
+                $ fromMaybe W.slotMinBound
+                $ W.slotAt' sp currentTime
```
I wasn't able to fully remove `slotAt'` yet in #1901, but we need to use the `TimeInterpreter` for this. Otherwise the calculation will be incorrect.

I suggest using `epochOf`, `ongoingSlotAt`, and running the resulting `Qry` using the `timeInterpreter` field from `networkLayer`. Happy to help if that doesn't work out for some reason.

jonathanknowles replied:
> I wasn't able to fully remove `slotAt'` yet in #1901, but we need to use the `TimeInterpreter` for this. Otherwise the calculation will be incorrect.

Thanks for letting me know. When are you planning to remove the `slotAt'` function? The `getNetworkInformation` endpoint also uses this function in a similar way to determine the _next_ epoch. If that function is updated to use the new API, then we would presumably have to do something similar for `joinStakePool`?

Anviking replied:
> If that function is updated to use the new API, then we would presumably have to do something similar for joinStakePool?

Yes, they should be similar.

Here you handle the `Nothing` case, where it means the current time is behind the genesis date, I think.

With the hard fork, there the query may (also) fail if the node is not in sync, if the current time is beyond the forecast range of the interpreter (think this is up to the next epoch, but not sure exactly)

jonathanknowles replied:
> > If that function is updated to use the new API, then we would presumably have to do something similar for joinStakePool?
> 
> Yes, they should be similar.
> 
> Here you handle the `Nothing` case, where it means the current time is behind the genesis date, I think.
> 
> With the hard fork, there the query may (also) fail if the node is not in sync, if the current time is beyond the forecast range of the interpreter (think this is up to the next epoch, but not sure exactly)

Hi @Anviking 

Thanks for your help with this!

I've (fingers crossed) fixed this with the following commits:

* 255210f8b5e6b58a7f4ce47ab83e26b954bc915b
* 7660579b295bd8ad78085da453c6c1c670f45bd5
* 18ef12a77a7a600738e52f85efe292de87168eef
* 23010f8ab79daa7f0ff86d0693b0f02b20c97c0a

jonathanknowles committed:
Report `ErrNoSuchPool` instead of `ErrPoolAlreadyRetired`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1913#issuecomment-662360793

jonathanknowles committed:
Add function `epochAt` to module `Slotting`.

jonathanknowles committed:
Add function `currentEpoch` to module `Slotting`.

jonathanknowles committed:
Add function `getCurrentEpoch` to API server layer.

jonathanknowles committed:
Pass through current epoch to `joinStakePool` in wallet layer.

rvl commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1877,6 +1899,12 @@ newtype ErrRejectedTip = ErrRejectedTip ApiNetworkTip
+    handler = \case
+        ErrUnableToDetermineCurrentEpoch ->
+            apiError err500 UnableToDetermineCurrentEpoch
+                "I'm unable to determine the current epoch."
```
Why would "I" be unable to do that?
Because the node is not synced?
What can the user do about it?

jonathanknowles replied:
> Why would "I" be unable to do that?
> Because the node is not synced?
> What can the user do about it?

I'm pretty sure there's not much the user can do about this.

I'm going to create another PR with an enhancement to this error, once I can determine the range of conditions under which it's not possible to determine the current epoch (which is derived from the current slot).

jonathanknowles replied:
Issue created here: https://github.com/input-output-hk/cardano-wallet/issues/1939

jonathanknowles committed:
Fix import order.
